src/lib/format/l20n/ast/parser.js

```diff
@@ -7,18 +7,20 @@ const MAX_PLACEABLES = 100;
 
 export default {
   parse: function(env, string, pos = false) {
-    this._source = string;
-    this._index = 0;
-    this._length = string.length;
-    this._curEntryStart = 0;
-    this._config = {
-      pos: pos
-    };
-
-    if (pos !== true) {
-      AST.Node.prototype.setPosition = function() {};
-    }
-    return this.getResource(pos);
+    AST.Node._config.pos = pos;
+
+    const parseContext = this.getNewParseContext(env, string, pos);
+
+    return parseContext.getResource();
+  },
+
+  getNewParseContext: function(env, string, pos = false) {
+    const parseContext = Object.create(this);
```

Comment:
Yes and no, depending on what you mean by _class_ given that this is JavaScript :)  With `const parseContext = Object.create(this);` you're essentially doing exactly what the `new` keyword does: creating a new object with a prototype set to another object.

By doing this in a more idiomatic way, the code might end up being more maintainable and also we might benefit from Gecko optimizations targeted at object instantiation.

So yeah, maybe the easiest way of achieving the same thing would be simply to have something like:

```javascript
export function parseString(str) {
  ...
  const pctx = new ParserContext(...);
  ...
}

function ParserContext(...) {
   this.index = 0;
   ...
}

ParserContext.prototype.getResource = function(...) { ... }
etc.
```

(or use es6 classes, it's the same thing.)